compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -327,19 +330,33 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
             let field_align = |f: &TyAndLayout<'_>| {
                 if let Some(pack) = pack { f.align.abi.min(pack) } else { f.align.abi }
             };
-            match kind {
-                StructKind::AlwaysSized | StructKind::MaybeUnsized => {
-                    optimizing.sort_by_key(|&x| {
-                        // Place ZSTs first to avoid "interesting offsets",
-                        // especially with only one or two non-ZST fields.
-                        let f = &fields[x as usize];
-                        (!f.is_zst(), cmp::Reverse(field_align(f)))
-                    });
-                }
-                StructKind::Prefixed(..) => {
-                    // Sort in ascending alignment so that the layout stay optimal
-                    // regardless of the prefix
-                    optimizing.sort_by_key(|&x| field_align(&fields[x as usize]));
+
+            // If `-Z randomize-layout` is set we can mess with field ordering to try
+            // and catch some people doing things they shouldn't be
+            if repr.can_randomize_struct_fields() {
+                // `ReprOptions.layout_seed` is a deterministic seed that we can use to
+                // randomize field ordering with
+                let mut rng = Xoshiro128StarStar::from_seed(repr.layout_seed);
+
+                // Shuffle the ordering of the fields
+                optimizing.shuffle(&mut rng);
+
+            // Otherwise we just leave things alone and actually optimize the struct fields
```

Comment:
"struct" is overly specific here, the fields could be for other things too.

Also, you can do both, if you let it sort the fields first, you could do this:
```rust
let alignment_groups = optimizing.group_by(|a, b| {
    field_align(&fields[a as usize]) == field_align(&fields[b as usize])
});
for group in alignment_groups {
    group.shuffle(&mut rng);
}
```
The benefit of this is that we can *always* do it, since it will never increase the type's size.

Reply:
I'd prefer `-Z randomize-layout` to land (and be available to people) before doing more interesting crater experiments with it, so I added a `FIXME` that we can go back to and play around with in future PRs